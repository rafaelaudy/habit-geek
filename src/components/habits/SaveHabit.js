import React, { useState } from "react";
import { useTranslation } from "react-i18next";

const SaveHabit = ({
  id,
  name: defaultName = "",
  type: defaultType = "",
  frequency: defaultFrequency = "",
  saveHabit,
  deleteHabit,
  onGoBack
}) => {
  const [name, setName] = useState(defaultName);
  const [type, setType] = useState(defaultType);
  const [frequency, setFrequency] = useState(defaultFrequency);
  const [validityChecked, setValidityChecked] = useState(false);
  const { t } = useTranslation();

  const saveAndClose = event => {
    event.preventDefault();
    setValidityChecked(true);

    if (event.target.checkValidity()) {
      saveHabit(id, name, type, frequency);
      onGoBack();
    }
  };

  const deleteHabitAndClose = () => {
    deleteHabit(id);
    onGoBack();
  };

  const typeOptions = t("save-habit-type-options", {
    returnObjects: true
  });

  return (
    <form
      onSubmit={saveAndClose}
      className={`mobile-size-container ${
        validityChecked ? "was-validated" : ""
      }`}
      noValidate
    >
      <h2>{t("save-habit-title")}</h2>
      <div className="mb-3">
        <label htmlFor="new-habit-name">{t("save-habit-name")}</label>
        <input
          id="new-habit-name"
          className="form-control"
          value={name}
          onChange={e => setName(e.target.value)}
          required
        ></input>
        <div className="invalid-feedback">
          {t("save-habit-name-validation")}
        </div>
      </div>
      <div className="mb-3">
        <label htmlFor="new-habit-type">{t("save-habit-type")}</label>
        <select
          id="new-habit-type"
          className="form-control"
          value={type}
          onChange={e => setType(e.target.value)}
          onBlur={e => setType(e.target.value)}
          required
        >
          <option value=""></option>
          {typeOptions.map(({ key, label }) => (
            <option key={`save-habit-type-option-${key}`} value={label}>
              {label}
            </option>
          ))}
        </select>
        <div className="invalid-feedback">
          {t("save-habit-type-validation")}
        </div>
      </div>
      <div className="mb-3">
        <label htmlFor="new-habit-frequency">{t("save-habit-frequency")}</label>
        <select
          id="new-habit-frequency"
          className="form-control"
          value={frequency}
          onChange={e => setFrequency(e.target.value)}
          onBlur={e => setFrequency(e.target.value)}
          required
        >
          <option value=""></option>
          <option value="1">1x</option>
          <option value="2">2x</option>
          <option value="3">3x</option>
          <option value="4">4x</option>
          <option value="5">5x</option>
          <option value="6">6x</option>
          <option value="7">7x</option>
        </select>
        <div className="invalid-feedback">
          {t("save-habit-frequency-validation")}
        </div>
      </div>
      <hr className="mb-4" />
      <button className="btn btn-primary btn-lg btn-block">
        {t("save-habit-save")}
      </button>
      {id ? (
        <button
          type="button"
          className="btn btn-danger btn-lg btn-block"
          onClick={deleteHabitAndClose}
        >
          {t("save-habit-delete")}
        </button>
      ) : null}
      <button
        type="button"
        className="btn btn-secondary btn-lg btn-block"
        onClick={onGoBack}
      >
        {t("save-habit-cancel")}
      </button>
    </form>
  );
};

export default SaveHabit;
