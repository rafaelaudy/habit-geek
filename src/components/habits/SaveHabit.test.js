import React from "react";
import { shallow } from "enzyme";
import SaveHabit from "./SaveHabit";

let fakeEvent;
let saveHabitMock;
let deleteHabitMock;
let onGoBackMock;

describe("SaveHabit component", () => {
  beforeEach(() => {
    saveHabitMock = jest.fn();
    deleteHabitMock = jest.fn();
    onGoBackMock = jest.fn();
    fakeEvent = {
      preventDefault: jest.fn(),
      target: { checkValidity: jest.fn().mockReturnValue(true) }
    };
  });

  it("Renders static elements", () => {
    const saveHabits = shallow(
      <SaveHabit name="name" frequency="frequency" type="1"></SaveHabit>
    );
    expect(saveHabits).toMatchSnapshot();
  });

  it("creates a habit", () => {
    const saveHabits = shallow(
      <SaveHabit saveHabit={saveHabitMock} onGoBack={onGoBackMock}></SaveHabit>
    );
    saveHabits
      .find("#new-habit-name")
      .simulate("change", { target: { value: "Read" } });
    saveHabits
      .find("#new-habit-type")
      .simulate("change", { target: { value: "Social" } });
    saveHabits
      .find("#new-habit-frequency")
      .simulate("change", { target: { value: "1x" } });
    saveHabits.find("form").simulate("submit", fakeEvent);
    expect(saveHabitMock).toHaveBeenCalledWith(
      undefined,
      "Read",
      "Social",
      "1x"
    );
    expect(onGoBackMock).toBeCalled();
  });

  it("updates a habit", () => {
    const saveHabits = shallow(
      <SaveHabit
        id="read"
        saveHabit={saveHabitMock}
        onGoBack={onGoBackMock}
      ></SaveHabit>
    );
    expect(saveHabits.find("h2").text()).toBe("save-habit-edit-title");
    saveHabits
      .find("#new-habit-name")
      .simulate("change", { target: { value: "Read" } });
    saveHabits
      .find("#new-habit-type")
      .simulate("change", { target: { value: "Social" } });
    saveHabits
      .find("#new-habit-frequency")
      .simulate("change", { target: { value: "1x" } });
    saveHabits.find("form").simulate("submit", fakeEvent);
    expect(saveHabitMock).toHaveBeenCalledWith("read", "Read", "Social", "1x");
    expect(onGoBackMock).toBeCalled();
  });

  it("validates forms", () => {
    const saveHabit = shallow(
      <SaveHabit saveHabit={saveHabitMock} onGoBack={onGoBackMock}></SaveHabit>
    );
    fakeEvent.target.checkValidity.mockReturnValueOnce(false);
    saveHabit.find("form").simulate("submit", fakeEvent);
    expect(saveHabit.find(".was-validated").length).toBe(1);
    expect(fakeEvent.preventDefault).toHaveBeenCalledTimes(1);
    expect(fakeEvent.target.checkValidity).toHaveBeenCalledTimes(1);
    expect(saveHabitMock).toHaveBeenCalledTimes(0);
    expect(onGoBackMock).toHaveBeenCalledTimes(0);
  });

  it("deletes a habit", () => {
    const saveHabits = shallow(
      <SaveHabit
        id="read"
        deleteHabit={deleteHabitMock}
        onGoBack={onGoBackMock}
      ></SaveHabit>
    );
    saveHabits.find(".btn-danger").simulate("click");
    expect(deleteHabitMock).toHaveBeenCalledWith("read");
    expect(onGoBackMock).toBeCalled();
  });

  it("selects change on blur for accessibility", () => {
    const saveHabits = shallow(
      <SaveHabit saveHabit={saveHabitMock} onGoBack={onGoBackMock}></SaveHabit>
    );
    saveHabits
      .find("#new-habit-type")
      .simulate("blur", { target: { value: "Social" } });
    saveHabits
      .find("#new-habit-frequency")
      .simulate("blur", { target: { value: "1x" } });
    saveHabits.find("form").simulate("submit", fakeEvent);
    expect(saveHabitMock).toHaveBeenCalledWith(undefined, "", "Social", "1x");
  });

  it("cancels habit creation", () => {
    const saveHabits = shallow(<SaveHabit onGoBack={onGoBackMock}></SaveHabit>);
    saveHabits.find(".btn-secondary").simulate("click");
    expect(onGoBackMock).toHaveBeenCalled();
  });
});
