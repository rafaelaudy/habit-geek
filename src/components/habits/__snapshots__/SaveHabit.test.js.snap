// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveHabit component Renders static elements 1`] = `
<form
  className="mobile-size-container "
  noValidate={true}
  onSubmit={[Function]}
>
  <h2>
    save-habit-new-title
  </h2>
  <div
    className="mb-3"
  >
    <label
      htmlFor="new-habit-name"
    >
      save-habit-name
    </label>
    <input
      className="form-control"
      id="new-habit-name"
      onChange={[Function]}
      required={true}
      value="name"
    />
    <div
      className="invalid-feedback"
    >
      save-habit-name-validation
    </div>
  </div>
  <div
    className="mb-3"
  >
    <label
      htmlFor="new-habit-type"
    >
      save-habit-type
    </label>
    <select
      className="form-control"
      id="new-habit-type"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      value="1"
    >
      <option
        value=""
      />
      <option
        key="save-habit-type-option-Health"
        value="Health"
      >
        Health
      </option>
      <option
        key="save-habit-type-option-Social"
        value="Social"
      >
        Social
      </option>
      <option
        key="save-habit-type-option-Career"
        value="Career"
      >
        Career
      </option>
      <option
        key="save-habit-type-option-Hobbies"
        value="Hobbies"
      >
        Hobbies
      </option>
    </select>
    <div
      className="invalid-feedback"
    >
      save-habit-type-validation
    </div>
  </div>
  <div
    className="mb-3"
  >
    <label
      htmlFor="new-habit-frequency"
    >
      save-habit-frequency
    </label>
    <select
      className="form-control"
      id="new-habit-frequency"
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      value="frequency"
    >
      <option
        value=""
      />
      <option
        value="1"
      >
        1x
      </option>
      <option
        value="2"
      >
        2x
      </option>
      <option
        value="3"
      >
        3x
      </option>
      <option
        value="4"
      >
        4x
      </option>
      <option
        value="5"
      >
        5x
      </option>
      <option
        value="6"
      >
        6x
      </option>
      <option
        value="7"
      >
        7x
      </option>
    </select>
    <div
      className="invalid-feedback"
    >
      save-habit-frequency-validation
    </div>
  </div>
  <hr
    className="mb-4"
  />
  <button
    className="btn btn-primary btn-lg btn-block"
  >
    save-habit-save
  </button>
  <button
    className="btn btn-secondary btn-lg btn-block"
    type="button"
  >
    save-habit-cancel
  </button>
</form>
`;
